### PR TITLE
Add text-only vote bot

### PR DESCRIPTION
**Pull request by wescpy**
- add text-only Apps Script/JavaScript vote bot
- add text-only Python App Engine vote bot
- update README

### CODE REVIEW

File: python/vote-text-bot/bot.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+        """
+        event = json.loads(self.request.body)
+        user = event['user']['displayName']
+        # Update vote when card button pressed
+        if event['type'] == 'CARD_CLICKED':
+            method = event['action']['actionMethodName']
+            # Create new vote when "NEW" button clicked
+            if method == 'newvote':
+                body = createMessage(user)
+            # Update card in-place when "+1" or "-1" button clicked
```
Comment (grant) [inline python/vote-text-bot/bot.py]: I'd put the comment closer to the logic.
```py
else:
    # Update card in-place when "+1" or "-1" button clicked
    delta = 1 ...
```

File: README.md [hunk 2]
```diff
@@ -25,6 +25,9 @@ following code samples:
     [card-formatted](https://developers.google.com/hangouts/chat/concepts/cards)
     response. This sample is an HTTP endpoint bot built upon
     [Google App Engine](https://cloud.google.com/appengine/).
+  - **Vote bot**: This bot demonstrates updating interactive cards by providing
+    a platform whereby users can vote on topics, such as who can do lunch today,
+    who wants to play ball at lunch, etc. Samples feature images or be text-only.
```
Comment (grant) [inline README.md]: Last sentence sounds weird. I think a `can` makes it more clear.
```
Samples can feature images or be text-only.
Samples feature images or can be text-only.
```

The first sentence sounds clearest to me.

File: python/vote-text-bot/bot.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+                                }
+                            }
+                        }
+                    }]
+                }]
+            }]
+        }]
+    }
+
+class MainPage(webapp2.RequestHandler):
```
Comment (grant) [inline python/vote-text-bot/bot.py]: `MainPage` doesn't really make sense for this server. Rename the server class.

Comment (grant) [inline python/vote-text-bot/bot.py]: Please comment this class.

File: python/vote-text-bot/bot.py [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+        elif event['type'] == 'MESSAGE':
+            body = createMessage(user)
+        else: # no response for ADDED_TO_SPACE or REMOVED_FROM_SPACE
+            return
+
+        # Send response back to Hangouts Chat
+        self.response.headers['Content-Type'] = 'application/json'
+        self.response.out.write(json.dumps(body))
+
+# start server loop
```
Comment (grant) [inline python/vote-text-bot/bot.py]: I don't think `start server loop` is completely accurate. I don't think there's explicit looping.

This snippet setups up a web server that listens to one path, `/`. Say that instead.

File: python/vote-text-bot/bot.py [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+'''
+Hangouts Chat simple text-only vote bot in Python App Engine
+
+@see developers.google.com/hangouts/chat/how-tos/cards-onclick
+'''
+
+import json
+import webapp2
+
+def createMessage(voter, vote_count=0, update=False):
```
Comment (grant) [inline python/vote-text-bot/bot.py]: Please follow [PEP 8 naming](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables): `create_message`.

File: apps-script/vote-text-bot/README.md [hunk 6]
```diff
@@ -0,0 +1,35 @@
+# Hangouts Chat text-only Vote bot sample
+
```
Comment (grant) [inline apps-script/vote-text-bot/README.md]: A picture of the bot would be great.

File: python/vote-text-bot/README.md [hunk 7]
```diff
@@ -0,0 +1,44 @@
+# Hangouts Chat text-only Vote bot sample
+
```
Comment (grant) [inline python/vote-text-bot/README.md]: A picture of the bot would be great.

File: python/vote-text-bot/bot.py (hunk 4 above)
Comment (wescpy) [reply python/vote-text-bot/bot.py]: DONE

File: README.md (hunk 2 above)
Comment (wescpy) [reply README.md]: DONE

File: apps-script/vote-text-bot/README.md (hunk 6 above)
Comment (wescpy) [reply apps-script/vote-text-bot/README.md]: DONE

File: apps-script/vote-text-bot/README.md [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+This code sample creates a Hangouts Chat simple text-only Vote bot
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/README.md]: DONE

File: apps-script/vote-text-bot/README.md [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+Script features tight integration with G Suite data and applications,
+developers can port it to Google Cloud Functions for a tighter
+integration with Google Cloud Platform if desired. The Vote bot can
+also be ported to Node.js for alternative hosting.
+
+## Run the sample
+
+Before you can run the sample, do the following:
+
+1. Read the [directions for setting up a Google Apps Script
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/README.md]: DONE

File: apps-script/vote-text-bot/README.md [hunk 10]
```diff
@@ -0,0 +1,35 @@
...
+  bot](http://developers.google.com/hangouts/chat/quickstart/apps-script-bot).
+1. Follow all of those directions _except_:
+  - Replace the template code with the Vote bot (vote-text-bot.js) in the editor.
+  - Name the project "Vote bot" and click **OK**.
+  - In the **Bot name** field, enter "Vote bot".
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/README.md]: DONE

File: apps-script/vote-text-bot/README.md [hunk 11]
```diff
@@ -0,0 +1,35 @@
...
+  - Under **Functionality**, select both checkboxes.
+  - Under **Connection settings**, select Apps Script project and paste the Deployment ID into the field.
+  - Under **Permissions**, select _Specific people and groups in your domain_, and enter your own email.
+1. Once you click the **Save changes** button, your bot will be published!
+1. Add the Vote bot to a chat room and see a new message from the bot with
+the vote card ready to accept its first vote!
+
+## Reference
+
+- Hangouts Chat developer [documentation](https://developers.google.com/hangouts/chat)
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/README.md]: DONE

File: apps-script/vote-text-bot/vote-text-bot.js [hunk 12]
```diff
@@ -0,0 +1,111 @@
...
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Hangouts Chat simple text-only vote bot in Google Apps Script
+ *
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/vote-text-bot.js]: DONE

File: apps-script/vote-text-bot/vote-text-bot.js [hunk 13]
```diff
@@ -0,0 +1,111 @@
...
+ * @see developers.google.com/hangouts/chat/how-tos/cards-onclick
+ */
+
+/**
+ * Create and return new interactive card or update existing one
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/vote-text-bot.js]: DONE

File: apps-script/vote-text-bot/vote-text-bot.js [hunk 14]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * @param {string} voter Google user name
+ * @param {number} voteCount Current vote count
+ * @param {boolean} update Update existing or create new card?
```
[the remark this answers is not in the document]
Comment (wescpy) [reply apps-script/vote-text-bot/vote-text-bot.js]: DONE

File: python/vote-text-bot/README.md (hunk 7 above)
Comment (wescpy) [reply python/vote-text-bot/README.md]: DONE

File: python/vote-text-bot/README.md [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+portable to Flask for App Engine or alternative hosting.
+
+## Run the sample
+
+Before you can run the sample, do the following:
+
+1. [Install](https://cloud.google.com/sdk/downloads) the Google Cloud SDK
+  if you haven't already and ensure the `gcloud` command works.
+1. Create a new [Google Cloud Platform](https://cloud.google.com) project
+  [here](https://console.cloud.google.com) or select an existing project,
```
[the remark this answers is not in the document]
Comment (wescpy) [reply python/vote-text-bot/README.md]: DONE

File: python/vote-text-bot/README.md [hunk 16]
```diff
@@ -0,0 +1,44 @@
...
+  and create an App Engine app (or using an existing one) and note the
+  _YOUR-APP_.appspot.com domain you get by default (customize as desired).
+1. Create a new folder and drop the files for this bot in it. Then from there,
+  [deploy](https://cloud.google.com/appengine/docs/standard/python/tools/uploadinganapp#deploying_an_app)
+  it to App Engine.
```
[the remark this answers is not in the document]
Comment (wescpy) [reply python/vote-text-bot/README.md]: DONE

File: python/vote-text-bot/README.md (hunk 15 above)
[the remark this answers is not in the document]
Comment (wescpy) [reply python/vote-text-bot/README.md]: DONE

File: python/vote-text-bot/README.md [hunk 17]
```diff
@@ -0,0 +1,44 @@
...
+1. Now that your bot is online, we need to
+  [publish](https://developers.google.com/hangouts/chat/how-tos/bots-publish)
+  it so Hangouts Chat can forward chat messages to it. Follow the instructions
```
[the remark this answers is not in the document]
Comment (wescpy) [reply python/vote-text-bot/README.md]: DONE

File: python/vote-text-bot/README.md [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+  on that page to enable the Hangouts Chat API. On the [Configuration
+  tab](https://console.developers.google.com/apis/api/chat.googleapis.com/hangouts-chat),
+  set the following values:
+  - Name the project "Vote bot" and click **OK**.
+  - In the **Bot name** field, enter "Vote bot".
```
[the remark this answers is not in the document]
Comment (wescpy) [reply python/vote-text-bot/README.md]: DONE

File: python/vote-text-bot/bot.py (hunk 5 above)
Comment (wescpy) [reply python/vote-text-bot/bot.py]: DONE

File: python/vote-text-bot/bot.py (hunk 3 above)
Comment (wescpy) [reply python/vote-text-bot/bot.py]: DONE

File: python/vote-text-bot/bot.py (hunk 1 above)
Comment (wescpy) [reply python/vote-text-bot/bot.py]: DONE

File: python/vote-text-bot/bot.py (hunk 3 above)
Comment (wescpy) [reply python/vote-text-bot/bot.py]: DONE

**Comment by grant** [comment]
LGTM!

**Pull request merged**